src/rules/array.jl

```diff
@@ -0,0 +1,56 @@
+#####
+##### `reshape`
+#####
+
+function rrule(::typeof(reshape), A::AbstractArray, dims::Tuple{Vararg{Int}})
+    return reshape(A, dims), (Rule(Ȳ->reshape(Ȳ, dims)), DNERule())
+end
+
+function rrule(::typeof(reshape), A::AbstractArray, dims::Int...)
+    Y, (rule, _) = rrule(reshape, A, dims)
+    return Y, (rule, fill(DNERule(), length(dims))...)
+end
+
+#####
+##### `hcat` (🐈)
+#####
+
+function rrule(::typeof(hcat), A::AbstractArray, Bs::AbstractArray...)
+    Y = hcat(A, Bs...)
+    Xs = (A, Bs...)
+    rules = ntuple(length(Bs) + 1) do i
+        l = mapreduce(j->size(Xs[j], 2), Base.add_sum, 1:i-1; init=0)
+        u = l + size(Xs[i], 2)
+        dim = u > l + 1 ? (l+1:u) : u
+        Rule(Ȳ->copy(selectdim(Ȳ, 2, dim)))
+    end
+    return Y, rules
+end
+
+#####
+##### `vcat`
+#####
+
+function rrule(::typeof(vcat), A::AbstractArray, Bs::AbstractArray...)
+    Y = vcat(A, Bs...)
+    n = size(A, 1)
+    ∂A = Rule(Ȳ->copy(selectdim(Ȳ, 1, 1:n)))
+    ∂Bs = ntuple(length(Bs)) do i
+        l = n + mapreduce(j->size(Bs[j], 1), Base.add_sum, 1:i-1; init=0)
+        u = l + size(Bs[i], 1)
+        Rule(Ȳ->copy(selectdim(Ȳ, 1, l+1:u)))
+    end
+    return Y, (∂A, ∂Bs...)
+end
+
+#####
+##### `fill`
+#####
+
+function rrule(::typeof(fill), value::Any, dims::Tuple{Vararg{Int}})
+    return fill(value, dims), (Rule(sum), DNERule())
+end
+
+function rrule(::typeof(fill), value::Any, dims::Int...)
+    return fill(value, dims), (Rule(sum), fill(DNERule(), length(dims))...)
```

Comment:
```suggestion
    return fill(value, dims), (Rule(sum), Fill(DNERule(), length(dims))...)
```

I thought we added a FillArrays dependency?
He haven't but we should?

Reply:
I wouldn't be pro- introducing this here until we've addressed #46 . Also, it doesn't look like `Fill`s have been optimised for splatting into a `Tuple` yet, so there's no real performance gain here.

Reply:
Oh, I missed that it was splatted
Fair enough

Reply:
We can make this compile time, if we use `ntuple`
```
julia> @btime (()->tuple(ntuple(_->Nothing(), 10)...))()
  0.027 ns (0 allocations: 0 bytes)
(nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing)

julia> @btime (()->tuple(fill(Nothing(), 10)...))()
  319.146 ns (1 allocation: 80 bytes)
(nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing, nothing)
```


```suggestion 
    return fill(value, dims), (Rule(sum), ntuple(_->DNERule(), length(dims))...)
```